### PR TITLE
feat: ics-feed calendar api-v2

### PR DESCRIPTION
**Pull request by ThyMinimalDev**
## What does this PR do?

POST calendars/ics-feed/save: This endpoint allows users to save an ICS feed to their calendar.
GET calendars/ics-feed/check: This endpoint checks the status of an existing ICS feed, verifying whether it has been properly synced or saved.

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A - I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
yarn test:e2e calendars.controller.e2e-spec.ts

### CODE REVIEW

File: apps/api/v2/src/ee/calendars/input/create-ics.input.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+export class IsICSUrlConstraint implements ValidatorConstraintInterface {
+  validate(url: unknown) {
+    if (typeof url !== "string") return false;
+
+    // Check if it's a valid URL and ends with .ics
+    try {
+      const urlObject = new URL(url);
+      return (
+        urlObject.protocol === "http:" ||
+        (urlObject.protocol === "https:" && urlObject.pathname.endsWith(".ics"))
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/input/create-ics.input.ts:23]: If it's "http" protocol we are not checking if it ends with ".ics" but we should.

File: apps/api/v2/src/ee/calendars/input/create-ics.input.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  @IsBoolean()
+  @ApiProperty({
+    example: false,
+    description: "Whether to allowing writing to the calendar or not",
+    type: "boolean",
+    required: false,
+    default: true,
+  })
+  @IsOptional()
+  read_only?: boolean = true;
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/input/create-ics.input.ts:51]: For consistency would change to camel case aka `readOnly`

File: apps/api/v2/src/ee/calendars/input/create-ics.input.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  defaultMessage() {
+    return "The URL must be a valid ICS URL (ending with .ics)";
+  }
+}
+
+export class CreateIcsFeedInputDto {
+  @IsArray()
+  @ArrayNotEmpty()
+  @IsNotEmpty({ each: true })
+  @Validate(IsICSUrlConstraint, { each: true }) // Apply the custom validator to each element in the array
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/input/create-ics.input.ts:39]: Need to add `@ApiProperty({` documentation here too - would add example URL and explain that it has to start with http / https and end with ics.

File: apps/api/v2/src/ee/calendars/input/create-ics.input.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  IsOptional,
+  Validate,
+  ValidatorConstraint,
+  ValidatorConstraintInterface,
+} from "class-validator";
+import { IsNotEmpty, IsArray } from "class-validator";
+
+// Custom constraint to validate ICS URLs
+@ValidatorConstraint({ async: false })
+export class IsICSUrlConstraint implements ValidatorConstraintInterface {
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/input/create-ics.input.ts:14]: Can we move this under DTO so file reads top to bottom?

File: apps/api/v2/src/ee/calendars/services/ics-feed.service.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  constructor(
+    private readonly calendarsService: CalendarsService,
+    private readonly credentialRepository: CredentialsRepository
+  ) {}
+
+  private logger = new Logger("IcsFeedService");
+
+  async save(
+    userId: number,
+    userEmail: string,
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/services/ics-feed.service.ts:22]: Can we make `save` accept either "userId" or "userEmail" but not both, and then if "userId" is passed we fetch the user and then pass down the user email manually - all this is to simplify developer experience.

File: apps/api/v2/src/ee/calendars/input/create-ics.output.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import { ApiProperty } from "@nestjs/swagger";
+import { Expose, Type } from "class-transformer";
+import { IsString, ValidateNested, IsEnum, IsInt, IsBoolean } from "class-validator";
+
+import { ERROR_STATUS, SUCCESS_STATUS } from "@calcom/platform-constants";
+
+export class CreateIcsFeedOutput {
+  @IsInt()
+  @Expose()
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/input/create-ics.output.ts:9]: For each property needs "ApiProperty" to have the output in the docs.

File: apps/api/v2/src/ee/calendars/controllers/calendars.controller.ts [hunk 7]
```diff
@@ -58,9 +62,26 @@ export class CalendarsController {
...
+    @GetUser("id") userId: number,
+    @GetUser("email") userEmail: string,
+    @Body() body: CreateIcsFeedInputDto
+  ): Promise<CreateIcsFeedOutputResponseDto> {
+    return await this.icsFeedService.save(userId, userEmail, body.urls, body.read_only);
+  }
+
+  @Get("/ics-feed/check")
+  @UseGuards(ApiAuthGuard)
+  async CheckIcsFeed(@GetUser("id") userId: number): Promise<ApiResponse> {
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/controllers/calendars.controller.ts:81]: first letter should be small `checkIcsFeed`

File: apps/api/v2/src/ee/calendars/controllers/calendars.controller.e2e-spec.ts [hunk 8]
```diff
@@ -192,6 +197,52 @@ describe("Platform Calendars Endpoints", () => {
       .expect(200);
   });
 
+  it(`/GET/v2/calendars/${ICS_CALENDAR}/save with access token should fail to create a new ics feed calendar credentials with invalid urls`, async () => {
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/controllers/calendars.controller.e2e-spec.ts:200]: Should be POST not GET in test "it" description

File: apps/api/v2/src/ee/calendars/controllers/calendars.controller.e2e-spec.ts [hunk 9]
```diff
@@ -192,6 +197,52 @@ describe("Platform Calendars Endpoints", () => {
...
+    };
+    await request(app.getHttpServer())
+      .post(`/v2/calendars/${ICS_CALENDAR}/save`)
+      .set("Authorization", `Bearer ${accessTokenSecret}`)
+      .set("Origin", CLIENT_REDIRECT_URI)
+      .send(body)
+      .expect(400);
+  });
+
+  it(`/GET/v2/calendars/${ICS_CALENDAR}/save with access token should create a new ics feed calendar credentials`, async () => {
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/controllers/calendars.controller.e2e-spec.ts:213]: Should be POST not GET in test "it" description

File: apps/api/v2/src/ee/calendars/services/ics-feed.service.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    urls: string[],
+    readonly = true
+  ): Promise<CreateIcsFeedOutputResponseDto> {
+    return await this.saveCalendarCredentials(userId, userEmail, urls, readonly);
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/services/ics-feed.service.ts:26]: We are calling `saveCalendarCredentials` only in `save` so why not move all of the content of `saveCalendarCredentials` into `save` ?

File: apps/api/v2/src/ee/calendars/services/ics-feed.service.ts [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  async check(userId: number): Promise<{ status: typeof SUCCESS_STATUS }> {
+    return await this.checkIfCalendarConnected(userId);
```
Comment (supalarry) [inline apps/api/v2/src/ee/calendars/services/ics-feed.service.ts:30]: We are calling checkIfCalendarConnected only in check so why not move all of the content of checkIfCalendarConnected into check ?

File: apps/api/v2/src/ee/calendars/services/ics-feed.service.ts (hunk 5 above)
Comment (ThyMinimalDev) [reply apps/api/v2/src/ee/calendars/services/ics-feed.service.ts:22]: we already fetch the user via @GetUser i'd rather not do it twice

**Review by supalarry — DISMISSED** [review]

**Review by ibex088 — DISMISSED** [review]

**Review by ibex088 — APPROVED** [review]

**Pull request merged**